### PR TITLE
fix(react-combobox): change useComboboxFilter hook to filter by option text

### LINKED ISSUE #31464: [Bug]: useComboboxFilter filters the value not the text
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
OS: Linux 5.15 Ubuntu 24.04 LTS 24.04 LTS (Noble Numbat)
    CPU: (16) x64 AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics
    Memory: 6.50 GB / 11.52 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
```

### Are you reporting Accessibility issue?

None

### Reproduction

import { Combobox, makeStyles, useComboboxFilter, useId, type ComboboxProps } from '@fluentui/react-components';
import * as React from 'react';

const useStyles = makeStyles({
    root: {
        // Stack the label above the field with a gap
        display: 'grid',
        gridTemplateRows: 'repeat(1fr)',
        justifyItems: 'start',
        gap: '2px',
        maxWidth: '400px',
    },
});

const options = [
    { children: 'Alligator', value: '1' },
    { children: 'Bee', value: '2' },
    { children: 'Bird', value: '3' },
    { children: 'Cheetah', disabled: true, value: '4' },
    { children: 'Dog', value: '5' },
    { children: 'Dolphin', value: '6' },
    { children: 'Ferret', value: '7' },
    { children: 'Firefly', value: '8' },
    { children: 'Fish', value: '9' },
    { children: 'Goat', value: '10' },
    { children: 'Horse', value: '11' },
    { children: 'Lion', value: '12' },
];

export const ComboboxIssue = () => {
    const comboId = useId();
    const styles = useStyles();

    const [query, setQuery] = React.useState<string>('');
    const children = useComboboxFilter(query, options, {
        noOptionsMessage: 'No animals match your search.',
    });
    const onOptionSelect: ComboboxProps['onOptionSelect'] = (e, data) => {
        setQuery(data.optionText ?? '');
    };

    return (
        <div className={styles.root}>
            <label id={comboId}>Best pet</label>
            <Combobox
                onOptionSelect={onOptionSelect}
                aria-labelledby={comboId}
                placeholder="Select an animal"
                onChange={(ev) => setQuery(ev.target.value)}
                value={query}
            >
                {children}
            </Combobox>
        </div>
    );
};

## Actual Behavior
`{ children: 'sometext', value: 'c9e0165d-6538-4300-aaf3-94c5b7f01fb7' }`
The combobox component displays the "children" property as selectable option. The useComboboxFilter hook does use the "value" property for filtering. If the "value" property is set to e.g. a guid (which is very likely in data driven applications) the user has to type the guid to get the combobox filtered .

## Expected Behavior
The filter is using the same property as displayed to the user for filtering (children).

### Logs

_No response_

### Requested priority

High

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by dmytrokirpa**
## Previous Behavior

The `useComboboxFilter` hook utilizes the value, not the text, to filter options.

## New Behavior

The `useComboboxFilter` hook utilizes the text to filter options, and if it's not string-like (html element), it defaults to the option value.

## Related Issue(s)

- Fixes #31464

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">31</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">40</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/Avatar-mount_base.html">608</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/Avatar-mount.html">654</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/Button-mount_base.html">308</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/Button-mount.html">295</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/Field-mount_base.html">1170</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/Field-mount.html">1152</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProvider-mount_base.html">706</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProvider-mount.html">696</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-mount_base.html">86</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-mount.html">87</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">31</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">40</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">80</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">72</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/MakeStyles-mount_base.html">855</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/MakeStyles-mount.html">882</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/Persona-mount_base.html">1783</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/Persona-mount.html">1678</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/SpinButton-mount_base.html">1350</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/SpinButton-mount.html">1400</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/SwatchPicker-mount_base.html">1632</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31747/perf-test-react-components/SwatchPicker-mount.html">1681</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

File: change/@fluentui-react-combobox-8e0ae73a-c593-485c-b6f0-a5f4e576a378.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-combobox-8e0ae73a-c593-485c-b6f0-a5f4e576a378.json:1]: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31747&commitId=fea48b47efd3dd5cc626b98dc8e52e9b2e0d38c4&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask.normal.chromium.png|3| Changed|

</details>


  <details><summary><b>Card Converged - Interactive</b> 6 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Interactive.appearance interactive - Filled - High Contrast.click.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Filled - High Contrast.hover.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Filled - High Contrast.normal.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Filled Alternative - High Contrast.click.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Filled Alternative - High Contrast.hover.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Filled Alternative - High Contrast.normal.chromium.png|0| Added|

</details>


  <details><summary><b>SwatchPicker Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|SwatchPicker Converged.Shape.default.chromium.png|0| Added|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Comment by JoseHerminioCollas** [comment]
The key values are still showing up in the display.

I can use an element as a `children` in the `options` array that I will provide to the `Combobox` as the following:
```
const options = [
  { children: <em>Alligator</em>, value: '1' },
  { children: 'Bee', value: '2' },

```
The `Option` with the `EM` tag is selected:
image

image

The value that is displayed is not the correct value
image

The expected behavior is that the `children` value is displayed not the `value` value.

**Comment by dmytrokirpa** [comment]
> The key values are still showing up in the display.
> 
> I can use an element as a `children` in the `options` array that I will provide to the `Combobox` as the following:
> 
> ```
> const options = [
>   { children: <em>Alligator</em>, value: '1' },
>   { children: 'Bee', value: '2' },
> ```
> 
> The `Option` with the `EM` tag is selected: image
> 
> image
> 
> The value that is displayed is not the correct value image
> 
> The expected behavior is that the `children` value is displayed not the `value` value.

Thank you, @JoseHerminioCollas, for highlighting this issue. The challenge is that HTML elements cannot be rendered within an input field. I have implemented some logic to accommodate this use-case, although I am uncertain of its value. Let's await the team's feedback.

**Review by layershifter — CHANGES_REQUESTED** [review]
Any direct manipulation of children should be avoid as it leads to unpredictable hard to explain results, for example:

```tsx
function Hello() {
  return <>Hello</>;
}

expect(childrenToString(<span>Hello</span>)).toEqual('Hello'); // ✅ 
expect(childrenToString(<Hello />)).toEqual('Hello'); // 🔴  won't pass as `Hello` does not have children
```

**Comment by dmytrokirpa** [comment]
> Any direct manipulation of children should be avoid as it leads to unpredictable hard to explain results, for example:
> 
> ```tsx
> function Hello() {
>   return <>Hello</>;
> }
> 
> expect(childrenToString(<span>Hello</span>)).toEqual('Hello'); // ✅ 
> expect(childrenToString(<Hello />)).toEqual('Hello'); // 🔴  won't pass as `Hello` does not have children
> ```

I concur that it may not be beneficial if it does not cover all scenarios.

@layershifter, what is your recommendation for this situation? Should we label the original issue as "won't do," or should we permit filtering by string-only children as in my original implementation [here](https://github.com/microsoft/fluentui/pull/31747/commits/75827a71ceb223c4a2de574ec4f578bc619c9622#diff-b27d23be71c760c92927df29d56390ca7c8fcaf7367369d00946623227c012caR36)?

**Comment by ling1726** [comment]
@dmytrokirpa  I don't think this feature is necessary. There is a `optionToText` function in the API that lets users determine what text is filtered on: here's an example https://stackblitz.com/edit/nhpsvt?file=src%2Fexample.tsx

**Pull request closed without merging**